### PR TITLE
linear algorithm for getting the top K elements

### PR DESCRIPTION
**Pull request by RaduBerinde**
Implementation of an algorithm to select the top K elements from a collection,
based on Hoare's randomized selection algorithm. Nate will experiment using
it in the sql layer when we have to sort but we also have a LIMIT.

The test takes ~0.08s.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5387)

### CODE REVIEW

File: util/topk.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"sort"
+)
+
+var vals []int
+
+// moveTopKToFront swaps elements in the range [start, end) so that all
+// elements in the range [start, k) are all <= than all elements in the range
+// [k, end).
+func moveTopKToFront(data sort.Interface, start, end, k int, rng *rand.Rand) {
+	//fmt.Println("Move", vals[start:end], k)
```
Comment (tschottdorf) [inline util/topk.go]: Debug code.

File: util/topk.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		}
+		if l == r {
+			break
+		}
+		data.Swap(l, r-1)
+		l++
+		r--
+	}
+	mid := l
+	//fmt.Println("After pivoting", vals[start:mid], vals[mid:end])
```
Comment (tschottdorf) [inline util/topk.go]: ditto

**Comment by nvanbenschoten** [comment]
Awesome! Thanks for pulling this all together into an easy to use util function. Once we push this through I'll experiment with swapping some of the sorting strategies to use this new approach. We should be able to get a nice boost to our algorithmic complexity over there.

---

Reviewed 2 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[util/topk.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBBbSqJQThLSosddFg:-KDBBbSrcPCVTXfMO7A-:-482004622) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L30)):*
It might be a good idea to replace `rng *rand.Rand` with `func rand(n int) int` so that we can play around with different sources of randomness (even if we initially call the function with `moveTopKToFront(..., r.Intn)`). I only say this because we really only need pseudo-randomness here, and if possible it would be nice to avoid the synchronization costs of `rand.Rand`. This would also allow us to play around with adding a "center affinity" to the randomness as we discussed in person.

On this note, does the pivot even need to be random? A lot of quicksort implementations I've seen just arbitrarily pick the first or last element to be the pivot (https://golang.org/src/sort/sort.go#L119).

---

*[util/topk.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBEbGdCdTkbFlRqbTv:-KDBEbGdCdTkbFlRqbTw:64197161) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L30)):*
Minor nit that you can ignore, but it looks like the stdlib uses the variable names `lo` and `high`. Maybe consider switching to be consistent. 

---

*[util/topk.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBFJm7ZR2FAL2sKyio:-KDBFJm7ZR2FAL2sKyip:-454593105) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L49)):*
I feel like this logic is repeating the `l < r` check a lot, which leads me to believe it can be simplified. For instance, I think it's safe to remove the check in the outer `for l < r {` loop and replace it with `for {` if we change the `if l == r {` to `if l < r {`.

---

*[util/topk.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBGRqbFRFRO4L7vy2W:-KDBGRqbFRFRO4L7vy2X:2914222) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L73)):*
Wouldn't this be `moveTopKToFront(data, mid, end, k - mid, rng)`? Maybe I'm missing something. I probably also have an off-by-one error there.

---

*[util/topk.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBEMqJvLpYY4TMH5V6:-KDBEMqJvLpYY4TMH5V7:110902617) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L106)):*
Just curious, what do you mean by "with very high probability"? Do you mean with high probability the number of calls will approach the worst-case upper bound?

---

*[util/topk_test.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBICT3yup1d_lZOqwi:-KDBICT3yup1d_lZOqwj:1117242858) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk_test.go#L49)):*
Do you mind adding a benchmark? It's always nice to have something to compare against when optimizing later.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-:-KDBE7ofQ3RdHT2_4TrZ:-1555821039)*

**Comment by nvanbenschoten** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[util/topk.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBEMqJvLpYY4TMH5V6:-KDBLtDZIXWgY0qhV8F3:1822553552) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L106)):*
Actually, is the worst case here `O(n)` or `O(n^2)`. It looks a lot like Quickselect.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387)*

**Comment by RaduBerinde** [comment]
TFTR!

---

Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[util/topk.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBBbSqJQThLSosddFg:-KDBJDebZWh2AvPZfwTf:1647888249) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L30)):*
Always picking the first or last is very bad as it can make things quadratic for the wrong input. The code you list does something more complicated (it looks at a bunch of values and takes the median). So yes, there are other ways, but random is a very good choice. 

I think there is some unwarranted aversion to using rand:
- there are no synchronization costs, we are using a "local" generator (only the global one uses a lock)
- the implementation doesn't look like it does anything slow (https://golang.org/src/math/rand/rng.go#L232)
- we are using O(log n) random numbers in an O(n) algorithm. It would have to be *really* slow in order to show up.

---

*[util/topk.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBGRqbFRFRO4L7vy2W:-KDBJ6QIym0-bcY3BiEs:1860137350) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L73)):*
K is not relative to start (start <= k <= end). I started the other way but things were easier like this.

---

*[util/topk.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBEMqJvLpYY4TMH5V6:-KDBL-hlX8lL64skb9P-:-1375592604) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L106)):*
It's a standard term for randomized algo analysis, it means that for some constant `A`, the probability that the running time will be higher than `An` is `<= 1/n`. See https://en.wikipedia.org/wiki/With_high_probability

This is much stronger than looking at the expected time and proves that you don't need to worry about the worst possible case (which yes is quadratic) because it is overwhelmingly improbable.

Yeah, it's exactly quickselect aka Hoare's selection algo.

---

*[util/topk_test.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBICT3yup1d_lZOqwi:-KDBMIuqbpsrAOpgFhKa:-1375655603) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk_test.go#L49)):*
Sure, will add.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-:-KDBNVNG2SOqHg8VHdsl:-485616692)*

**Comment by nvanbenschoten** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[util/topk.go, line 30 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBBbSqJQThLSosddFg:-KDBRDAZ3qprRAR5EmWf:1960123182) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L30)):*
SGTM. I didn't realize Go made the distinction between `source`s and `lockedSource`s. 

---

*[util/topk.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBGRqbFRFRO4L7vy2W:-KDBQjVSL7fuED7z23hm:1254821226) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L73)):*
Ahh ok got it. Might be worth a comment to avoid future confusion.

---

*[util/topk.go, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBEMqJvLpYY4TMH5V6:-KDBQtiTKVsz881Dk88n:1419519330) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L106)):*
That makes a lot more sense. Thanks for the explanation. Do you mind citing Hoare's selection algo in the comments somewhere here?

---

*[util/topk_test.go, line 49 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBICT3yup1d_lZOqwi:-KDBRJMKhId0JUpA-cLw:-1246867683) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk_test.go#L49)):*
Thanks, and I would think we would want the benchmark input to be nondeterministic so we're not optimizing a specific case.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387)*

**Comment by RaduBerinde** [comment]
Nits fixed, added benchmarks (1.2ms for 100k values).

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/topk.go, line 73 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBGRqbFRFRO4L7vy2W:-KDBTw70WBgStf5cpxAJ:788159517) ([raw file](https://github.com/cockroachdb/cockroach/blob/79c0fd3ebcabdd156a37c794bcf9713a7a4549ac/util/topk.go#L73)):*
The function comment explains this very precisely.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-:-KDBVxTkXRhtHe6nIiJ8:1232466220)*

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r3.
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[util/topk_test.go, line 91 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBY2cFTiCQgjABCT3N:-KDBY2cGkfefcmwoV658:-1469467861) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6668c430d48363feb8aa81b0e41da170c1ce949/util/topk_test.go#L91)):*
Do we need this copy if we're copying below as well?

---

*[util/topk_test.go, line 94 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-KDBY6Mzbic80yiZ6EjF:-KDBY6Mzbic80yiZ6EjG:-1559973256) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6668c430d48363feb8aa81b0e41da170c1ce949/util/topk_test.go#L94)):*
Can we move this before ResetTimer?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5387#-:-KDBYk73rXCNa3Ht4HWy:1417088721)*

**Pull request merged**